test/unit/math/rev/mat/functor/algebra_solver_test.cpp

```diff
@@ -0,0 +1,658 @@
+  }
+};
+
+TEST(MathMatrix, simple_Eq) {
+  using stan::math::algebra_solver;
+  using stan::math::sum;
+  using stan::math::var;
+
+  int n_x = 2, n_y = 3;
+  for (int k = 0; k < n_x; k++) {
+
+    Eigen::VectorXd x(n_x);
+    x << 1, 1;  // initial guess
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y(n_y);
+    y << 5, 4, 2;
+    std::vector<double> dummy_dat(0);
+    std::vector<int> dummy_dat_int(0);
+    Eigen::Matrix<var, Eigen::Dynamic, 1> theta;
+
+    theta = algebra_solver(simpleEq_functor(),
+                           x, y, dummy_dat, dummy_dat_int);
+
+    EXPECT_EQ(20, theta(0));
+    EXPECT_EQ(2, theta(1));
+
+    Eigen::MatrixXd J(n_x, n_y);
+    J << 4, 5, 0, 0, 0, 1;
+
+    AVEC y_vec = createAVEC(y(0), y(1), y(2));
+    VEC g;
+    theta(k).grad(y_vec, g);
+
+    for (int i = 0; i < n_y; i++)
+      EXPECT_EQ(J(k, i), g[i]);
+  }
+}
+
+TEST(MathMatrix, simple_Eq_dbl) {
+  using stan::math::algebra_solver;
+
+ Eigen::VectorXd x(2);
+ x << 1, 1;  // initial guess
+ Eigen::VectorXd y(3);
+ y << 5, 4, 2;
+ std::vector<double> dummy_dat(0);
+ std::vector<int> dummy_dat_int(0);
+ Eigen::VectorXd theta;
+
+ theta = algebra_solver(simpleEq_functor(),
+                        x, y, dummy_dat, dummy_dat_int);
+
+ EXPECT_EQ(20, theta(0));
+ EXPECT_EQ(2, theta(1));
+}
+
+TEST(MathMatrix, simple_Eq_tuned) {
+  using stan::math::algebra_solver;
+  using stan::math::sum;
+  using stan::math::var;
+
+  int n_x = 2, n_y = 3;
+  for (int k = 0; k < n_x; k++) {
+
+    Eigen::VectorXd x(n_x);
+    x << 1, 1;  // initial guess
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y(n_y);
+    y << 5, 4, 2;
+    std::vector<double> dummy_dat(0);
+    std::vector<int> dummy_dat_int(0);
+
+    Eigen::Matrix<var, Eigen::Dynamic, 1> theta;
+    double xtol = 1e-6, ftol = 1e-6;
+    int maxfev = 1e+4;
+
+    theta = algebra_solver(simpleEq_functor(),
+                           x, y, dummy_dat, dummy_dat_int,
+                           0, xtol, ftol, maxfev);
+
+    EXPECT_EQ(20, theta(0));
+    EXPECT_EQ(2, theta(1));
+
+    Eigen::MatrixXd J(n_x, n_y);
+    J << 4, 5, 0, 0, 0, 1;
+
+    AVEC y_vec = createAVEC(y(0), y(1), y(2));
+    VEC g;
+    theta(k).grad(y_vec, g);
+
+    for (int i = 0; i < n_y; i++)
+      EXPECT_EQ(J(k, i), g[i]);
+  }
+}
+
+TEST(MathMatrix, simple_Eq_tuned_dbl) {
+  using stan::math::algebra_solver;
+  using stan::math::var;
+
+  int n_x = 2, n_y = 3;
+
+  Eigen::VectorXd x(n_x);
+  x << 1, 1;  // initial guess
+  Eigen::VectorXd y(n_y);
+  y << 5, 4, 2;
+  std::vector<double> dummy_dat(0);
+  std::vector<int> dummy_dat_int(0);
+
+  Eigen::VectorXd theta;
+  double xtol = 1e-6, ftol = 1e-6;
+  int maxfev = 1e+4;
+
+  theta = algebra_solver(simpleEq_functor(),
+                         x, y, dummy_dat, dummy_dat_int,
+                         0, xtol, ftol, maxfev);
+
+  EXPECT_EQ(20, theta(0));
+  EXPECT_EQ(2, theta(1));
+}
+
+///////////////////////////////////////////////////////////////////////////////
```

Comment:
If you need this kind of separator, think about including extra function and class defs in a separate file and including it.  Also, should be `nonlinear_eq_functor`, but we're not so picky with test code, so feel free to leave it.  Also, `scalar` should probably be `scalar_t` to indicate it's a typedef (that's a standard convention, but not part of Google code's style if I recall).